classes/class-wc-connect-paypal-ec.php

```diff
@@ -0,0 +1,250 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+if ( ! class_exists( 'WC_Connect_PayPal_EC' ) ) {
+
+	/**
+	 * Modify PPEC plugin behavior to facilitate proxying and authenticating requests via server
+	 */
+	class WC_Connect_PayPal_EC {
+
+		/**
+		 * @var WC_Connect_Nux
+		 */
+		private $nux;
+
+		public function __construct( WC_Connect_Nux $nux ) {
+			$this->nux = $nux;
+		}
+
+		public function init() {
+			if ( ! function_exists( 'wc_gateway_ppec' ) ) {
+				return;
+			}
+
+			$this->initialize_settings();
+
+			$settings = wc_gateway_ppec()->settings;
+			add_filter( 'woocommerce_paypal_express_checkout_settings', array( $this, 'paypal_ec_settings_meta' ) );
+			add_action( 'load-woocommerce_page_wc-settings', array( $this, 'maybe_set_reroute_requests' ) );
+
+			if ( 'yes' === $settings->reroute_requests ) {
+				// If empty, populate Sandbox and Live API Subject values with provided email
+				if (
+					empty( $settings->sandbox_api_subject ) &&
+					empty( $settings->sandbox_api_username ) &&
+					empty( $settings->api_username )
+				) {
+					$email = isset( $settings->email ) ? $settings->email : $settings->api_subject;
+					$settings->api_subject = $email;
+					$settings->sandbox_api_subject = $email;
+					$settings->save();
+				}
+
+				$username = $settings->get_active_api_credentials()->get_username();
+				$subject  = $settings->get_active_api_credentials()->get_subject();
+
+				if ( empty( $username ) && ! empty( $subject ) ) {
+					// Reroute Express Checkout requests from the PPEC extension via WCS to pick up API credentials
+					add_filter( 'woocommerce_paypal_express_checkout_request_body', array( $this, 'request_body' ) );
+
+					add_filter( 'option_woocommerce_ppec_paypal_settings', array( $this, 'paypal_ec_settings' ) );
+					add_filter( 'woocommerce_payment_gateway_supports', array( $this, 'supports' ), 10, 3 );
+
+					add_filter( 'pre_option_wc_gateway_ppce_prompt_to_connect', '__return_empty_string' );
+					if ( 'sandbox' !== $settings->environment ) {
+						add_action( 'woocommerce_order_status_on-hold', array( $this, 'maybe_set_banner' ) );
+						add_action( 'woocommerce_payment_complete', array( $this, 'maybe_set_banner' ) );
+						add_action( 'current_screen', array( $this, 'maybe_init_banner' ) );
+					}
+				}
+			}
+		}
+
+		public function request_body( $body ) {
+			$methods_to_proxy = array( 'SetExpressCheckout', 'GetExpressCheckoutDetails', 'DoExpressCheckoutPayment' );
+			if ( in_array( $body['METHOD'], $methods_to_proxy ) ) {
+				add_filter( 'woocommerce_paypal_express_checkout_request_endpoint', array( $this, 'request_endpoint' ), 10, 2 );
+			} else {
+				remove_filter( 'woocommerce_paypal_express_checkout_request_endpoint', array( $this, 'request_endpoint' ), 10, 2 );
```

Comment:
An alternative to removing the filter here is to do it in the `request_endpoint()` method so it's only ever executed once per attachment.

Reply:
Actually, we can't rely on this (and in fact it doesn't work) because the endpoint is retrieved multiple times per request: https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/includes/class-wc-gateway-ppec-client.php#L131-L133

Open to other suggestions for improving this though.